File: python/107_Binary_Tree_Level_Order_Traversal_II.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param {TreeNode} root
    # @return {integer[][]}
    def levelOrderBottom(self, root):
        if not root:
            return []
        queue=[root]
        res=[]
        while queue:
            nextLevel=[]
            curLine=[]
            for node in queue:
                curLine.append(node.val)
                if node.left:
                    nextLevel.append(node.left)
                if node.right:
                    nextLevel.append(node.right)
            queue=nextLevel
            res.append(curLine)
        return res[::-1]
                

File: Java/107_Binary_Tree_Level_Order_Traversal_II.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
// ~ 10ms
public class Solution {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        LinkedList<List<Integer>> res = new LinkedList<List<Integer>>();
        if(root==null)return res;
        ArrayList<TreeNode> queue = new ArrayList<>(), next;
        ArrayList<Integer> k;
        queue.add(root);
        while(queue.size()>0){
            k = new ArrayList<>();
            next = new ArrayList<>();
            for(TreeNode n:queue){
                k.add(n.val);
                if(n.left!=null)next.add(n.left);
                if(n.right!=null)next.add(n.right);
            }
            res.addFirst(k);
            queue=next;
        }
        return res;
    }
}